components/fraud_proof.md

```diff
@@ -0,0 +1,64 @@
+# Fraud Proof Algorithm for Simple Sequencer
+This is a WIP -- mostly just a few notes that I'm jotting down to not forget.
+
+## Requirements
+1. Not all blockhashes are posted up front.
+    - Blockhashes should be posted roughly every 100 L1 blocks.
+2. Only one block is evaluated during the fraud proof.
+
+## Fraud Proof Structure
+
+Fraud proofs are initialized with some constant size input. In our case this is:
+
+```python
+def init_fraud_proof(latest_blockhash: bytes, invalid_epoch_commitment: int):
+    ''' Invalid epoch commitment is the index of the invalid blockhash '''
+    # ...
+```
+
+What makes rollups special is that all of the preimages to the `latest_blockhash` are known. This means that from within the fraud proof VM, all blockhash preimages (including the state root, pre blockhash, etc) can be queried using the `preimage oracle`. Fundamentally this means that all historical blocks, events, transactions, and state can are accessible within the fraud proof VM.
+
+### Preimage Oracle
+
+The fraud proof VM has access to the following function (called the "preimage oracle"):
+
+```python
+def get_preimage(hash: bytes) -> bytes:
+    # The preimage is looked up from a local database held by the fraud prover
+    preimage = fetch_preimage(hash)
+    assert keccak256(preimage) == hash
+    return preimage
+```
+
+This seemingly simple functionality has extreme consequences. All publically available preimages can be operated on from within the fraud proof -- a pretty amazing property.
+
+### Step 1: Make available the blockhash disagreement
+- Requirement (1) reminder: the proposer does not submit all intermediate blockhashes.
+- Requirement (2) reminder: The fraud proof VM must only execute a single blockhash.
+
+The only way for us to satisfy these two requirements is to break our fraud proof into two distinct steps. 1) make all the blockhashes that were skipped available, and then 2) use the blockhashes & other Ethereum chain data inside the fraud proof VM to determine the invalid blockhash and execute the transition.
+
+So for step 1 all we need to do is force the proposer to post and commit to all intermediate blockhashes.
+
+### Step 2: Determine first invalid execution step
+
+The fraud proof function is executed in a Fraud Proof VM with access to the `preimage oracle`. The function is roughly:
+
+```python
+# The fraud proof function determines whether or not the supplied blockhash is
+# incorrect. This is not executed on chain but instead inside of the Fraud Proof VM
+# and then a single step of the fraud proof VM is executed on chain.
+def is_invalid_blockhash(latest_l1_blockhash: str, proposed_l2_blockhash: int)->bool:
```

Comment:
```
// this is what we store in the proposal manager
// epoch number ->  deterministic l2 state -> bytes32 (proposal hash)
// key->key->value
// ^^ this prevents collisions because you can make sure that there are no other proposals proposing the same state to avoid bad collisions.

// proposal type
type Proposal {
  l1BlockHash, proposerID
}

// this is run in the FPVM
fraud_proof(epoch_number, l2_state, proposal.l1BlockHash):
```

Reply:
> I admit to having been confused for a few minutes, because is_invalid_blockhash is basically the function on which we are going to play a challenge game on - right?

Yes sorry that function is evaluated entirely off-chain - I can see why that was confusing. Sorry about that!

> I'm also a bit confused, because I thought the function on which we play the challenge game on was "simply" going to be some sort of create_block_and_return_blockhash that uses geth. Feels like I'm missing something.

So it is totally reasonable to play the challenge game and return a boolean, or play the game and return a hash and afterwards compare that hash. Both are totally valid & I'm not opinionated on which to do. I don't *think* that you're missing anything there.